File: 0081-search-in-rotated-sorted-array-ii/0081-search-in-rotated-sorted-array-ii.cpp
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int n=nums.size();
        int low=0,high=n-1;
        while(low<=high){
            int mid=low+(high-low)/2;
            if(nums[mid]==target)
                return true;
            // specific...
           if(nums[low]==nums[mid] && nums[mid]==nums[high]){
                low=low+1;
                high=high-1;
                continue;
            }
            // search only in sorted part...
            if(nums[low]<=nums[mid]){
                if(nums[low]<=target && nums[mid]>=target)
                {
                    high=mid-1;
                }
                else
                    low=mid+1;
            }
            else
            {
                // right is sorted...
                if(nums[mid]<=target && nums[high]>=target){
                    low=mid+1;
                }
                else
                    high=mid-1;
            }
        }
        return false;
    }
};

File: 0081-search-in-rotated-sorted-array-ii/0081-search-in-rotated-sorted-array-ii.java
class Solution {
    public boolean search(int[] nums, int target) {
        int low = 0;
        int high = nums.length-1;
        while (low <= high) {
            int mid = (low + high) >> 1;
            if (nums[mid] == target)
                return true;
            if (nums[low] == nums[mid] && nums[mid] == nums[high]) {
                low = low + 1;
                high = high - 1;
                continue;
            }
            if (nums[low] <= nums[mid]) {

                if (nums[low] <= target && nums[mid] > target)// left is sorted
                {
                    high = mid - 1;
                } else
                    low = mid + 1;
            } else { // right is sorted
                if (nums[mid] < target && target <= nums[high]) {
                    low = mid + 1;
                } else
                    high = mid - 1;
            }
        }
        return false;
    }
}